algorithms/src/snark/marlin/marlin.rs

```diff
@@ -559,33 +629,104 @@ where
+            }
         }
 
+        // collect values into structures for our calculations
+        let mut max_constraint_domain = None;
+        let mut max_non_zero_domain = None;
+        let mut public_inputs = BTreeMap::new();
+        let mut padded_public_vec = Vec::with_capacity(keys_to_inputs.len());
+        let mut inputs_and_batch_sizes = BTreeMap::new();
+        let mut input_domains = BTreeMap::new();
+        let mut circuit_infos = BTreeMap::new();
+        let mut vks = Vec::with_capacity(keys_to_inputs.len());
+        let mut circuit_ids = Vec::with_capacity(keys_to_inputs.len());
+        for (vk, public_inputs_i) in keys_to_inputs.iter() {
+            vks.push(&vk.orig_vk.verifier_key);
+
+            let constraint_domains =
+                AHPForR1CS::<_, MM>::max_constraint_domain(&vk.orig_vk.circuit_info, max_constraint_domain)?;
+            max_constraint_domain = constraint_domains.max_constraint_domain;
+            let non_zero_domains =
+                AHPForR1CS::<_, MM>::max_non_zero_domain(&vk.orig_vk.circuit_info, max_non_zero_domain)?;
+            max_non_zero_domain = non_zero_domains.max_non_zero_domain;
+
+            let input_domain = EvaluationDomain::<E::Fr>::new(vk.orig_vk.circuit_info.num_public_inputs).unwrap();
+            input_domains.insert(vk.orig_vk.id, input_domain);
+
+            let (padded_public_inputs_i, parsed_public_inputs_i): (Vec<_>, Vec<_>) = {
+                public_inputs_i
+                    .iter()
+                    .map(|input| {
+                        let input = input.borrow().to_field_elements().unwrap();
+                        let mut new_input = vec![E::Fr::one()];
+                        new_input.extend_from_slice(&input);
+                        new_input.resize(input.len().max(input_domain.size()), E::Fr::zero());
+                        if cfg!(debug_assertions) {
+                            println!("Number of padded public variables: {}", new_input.len());
+                        }
+                        let unformatted = prover::ConstraintSystem::unformat_public_input(&new_input);
+                        (new_input, unformatted)
+                    })
+                    .unzip()
+            };
+            let circuit_id = vk.orig_vk.id;
+            public_inputs.insert(circuit_id, parsed_public_inputs_i);
+            padded_public_vec.push(padded_public_inputs_i);
+            circuit_infos.insert(circuit_id, &vk.orig_vk.circuit_info);
+            circuit_ids.push(circuit_id);
+        }
+        for (i, (vk, &batch_size)) in keys_to_inputs.keys().zip(batch_sizes.values()).enumerate() {
+            inputs_and_batch_sizes.insert(vk.orig_vk.id, (batch_size, padded_public_vec[i].as_slice()));
+        }
+
+        let verifier_key = VerifierUnionKey::<E>::union(vks);
+
+        let circuit_commitments =
+            keys_to_inputs.iter().map(|(vk, _)| &vk.orig_vk.circuit_commitments).collect::<Vec<_>>();
```

Comment:
same as https://github.com/AleoHQ/snarkVM/pull/1382/files#r1180229763

Reply:
in this case `circuit_commitments` is used once again [later on](https://github.com/AleoHQ/snarkVM/pull/1382/files#diff-9ce6b26e022266336b7600e2c82c3b4c76b7ef712e054ff9c1abc40f79daf72bR857), but it can just be re-created as an iterator again, consuming `keys_to_inputs`

Reply:
cef8f291b